samples/2_Cookbook/13_occupancy/occupancy.cpp

```diff
@@ -0,0 +1,167 @@
+IMPLIED, INNCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANNY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER INN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR INN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+*/
+
+#include "hip/hip_runtime.h"
+#include <iostream>
+#define NUM 1000000
+
+const unsigned threadsperblock = 32;
+const unsigned blocks = (NUM/threadsperblock)+1;
+
+uint32_t gridSize = 0;
+uint32_t blockSize = 0;
+
+#define HIP_CHECK(status)                                                                          \
+    if (status != hipSuccess) {                                                                    \
+        std::cout << "Got Status: " << status << " at Line: " << __LINE__ << std::endl;            \
+        exit(0);                                                                                   \
+    }
+
+// Device (Kernel) function
+__global__ void multiply(float* C, float* A, float* B, int N){
+      
+    int tx = blockDim.x*blockIdx.x+threadIdx.x;
+    
+    if (tx < N){
+	C[tx] = A[tx] * B[tx];
+    }
+}
+// CPU implementation
+void multiplyCPU(float* C, float* A, float* B, int N){
+    
+    for(unsigned int i=0; i<N; i++){     
+        C[i] = A[i] * B[i];      
+    }
+}
+
+void launchKernel(float* C, float* A, float* B, int N, bool manual){
+     
+     hipDeviceProp_t devProp;
+     HIP_CHECK(hipGetDeviceProperties(&devProp, 0));
+
+     hipEvent_t start, stop;
+     HIP_CHECK(hipEventCreate(&start));
+     HIP_CHECK(hipEventCreate(&stop));
+     float eventMs = 1.0f;
+     int activeWarps, maxWarps;
+      
+     if (manual){
+	blockSize = threadsperblock; 
+     	gridSize  = blocks;
+        std::cout << std::endl << "Manual Configuration with block size " << blockSize << std::endl;
+     }
+     else{
+     	hipOccupancyMaxPotentialBlockSize(&gridSize, &blockSize, multiply, 0, 0);
+	std::cout << std::endl << "Automatic Configuation based on hipOccupancyMaxPotentialBlockSize " << std::endl;
+	std::cout << "Suggested blocksize is " << blockSize << ", Minimum gridsize is " << gridSize << std::endl; 
+     }
+
+     // Record the start event
+     HIP_CHECK(hipEventRecord(start, NULL));  
+
+     // Launching the Kernel from Host
+     hipLaunchKernelGGL(multiply, dim3(gridSize), dim3(blockSize), 0, 0, C, A, B, NUM);
+     
+     // Record the stop event
+     HIP_CHECK(hipEventRecord(stop, NULL));
+     HIP_CHECK(hipEventSynchronize(stop));
+         
+     HIP_CHECK(hipEventElapsedTime(&eventMs, start, stop));
+     printf("kernel Execution time = %6.3fms\n", eventMs);
+
+     //Calculate Occupancy
+     uint32_t numBlock = 0;
+     hipOccupancyMaxActiveBlocksPerMultiprocessor(&numBlock, multiply, blockSize, 0);
+     
+     if(devProp.warpSize){
```

Comment:
warpSize does not impact the final theoretical occupancy calculation since it cancels out. You could simplify this if you wanted to just do (numBlock * blockSize / devProp.maxThreadsPerMultiProcessor). You can check for maxThreadsPerMultiProcessor being 0.